### ISSUE TITLE
Cookies are not being persisted

### ISSUE THREAD

**Issue by avindra** [guzzle/guzzle#1153]
Hi,

Using Guzzle 6.0.1. PHP 5.6.10. Consider the following example:

```php
		$jar = new \GuzzleHttp\Cookie\FileCookieJar("/tmp/somecookie");
		$client = new \GuzzleHttp\Client([
			"base_uri" => "http://api.example.com/v1/",
			"cookies"=> $jar
		]);
```

I would expect the file to have the cookie data saved, but upon shutting down the script / page, nothing is saved and the file ```/tmp/somecookie``` just has:

```javascript
    []
```

**Comment by mtdowling** [comment]
Hm. Session cookies are not persisted. Maybe you're working with session cookies that have no expires?

**Comment by avindra** [comment]
Not sure, but these are the curl options we were using in the previous spaghetti code (which did persist sessions):

```php
curl_setopt($this->curl, CURLOPT_COOKIEJAR,  $this->ckfile);
curl_setopt($this->curl, CURLOPT_COOKIESESSION, false );
curl_setopt($this->curl, CURLOPT_COOKIEFILE, $this->ckfile);
curl_setopt($this->curl, CURLOPT_RETURNTRANSFER, 1);
curl_setopt($this->curl, CURLOPT_FOLLOWLOCATION, 1);
```

**Comment by mtdowling** [comment]
Guzzle only persists non session cookies because each client is considered a session (similar to how browsers work). Can you check if you are using session cookies to see if that's the issue? You can see the headers sent over the wire using the "debug" option.

**Comment by Kofel** [comment]
I have similar issue. Guzzle is being ignore following cookie:
"Set-Cookie: JSESSIONID=64DD16D8A9DB617C2B914D62FE346159; Path=/registry"

**Comment by mtdowling** [comment]
Yeah, that's a session cookie which Guzzle does not persist as those are only meant to last for a single session (e.g., the life of a client).

What is the use case for persisting session cookies across clients? Can you just reuse the same client?

**Comment by avindra** [comment]
If a client is constructed every time there is a service request, then the session ID will be regenerated every time, which will create many wasteful sessions on the target service.

Is there a way to re-use a client after it has been destroyed? (My guess is no)

**Comment by mtdowling** [comment]
> Is there a way to re-use a client after it has been destroyed?

No

I'd be willing to merge a PR to optionally persist session cookies. For example the FileCookieJar constructor could have a new optional argument set to false by default that controls whether or not cookies with no expires value are persisted. See: https://github.com/guzzle/guzzle/blob/master/src/Cookie/FileCookieJar.php#L47

The same thing could be done for the SessionCookieJar: https://github.com/guzzle/guzzle/blob/master/src/Cookie/SessionCookieJar.php#L39

**Comment by Kofel** [comment]
@avindra  I faced with that problem, and my solution is to extend FileCookieJar and just reimplement save method with avoiding checking on session cookies.

@mtdowling + it would be great with this

**Comment by EAnushan** [comment]
The constructor option seems like it would be useful. Running into the same issue as others here. I'm making REST calls to an API and would like to preserve the Session Cookies returned by the service between page refreshes. Right now, I have to extend the SessionCookieJar myself and add the functionality.

**Comment by kmorozov** [comment]
I made a pull request to add a "save session cookies" parameter to the SessionCookieJar and FileCookieJar. Please Take a look:
https://github.com/guzzle/guzzle/pull/1195

**Comment by mtdowling** [comment]
Fixed in #1195

**Comment by avindra** [comment]
Works beautifully. Thank you @kmorozov  and @mtdowling
